### PR TITLE
Add support for directoryName subjectAltName

### PR DESCRIPTION
**Pull request by AndrzejKurek**
Continued work from https://github.com/Mbed-TLS/mbedtls/pull/731 in a similar fashion to https://github.com/Mbed-TLS/mbedtls/pull/6847, this PR adds support for directoryName subtype of subjectAltName.
https://mbed-tls.readthedocs.io/en/latest/kb/development/adding_new_subject_alternative_name_support/ was also roughly followed.

## Gatekeeper checklist

- [x] **changelog** provided
- [x] **backport** not required, this is a new feature
- [x] **tests** provided

### CODE REVIEW

File: include/mbedtls/x509_crt.h
Comment (mprse) [inline include/mbedtls/x509_crt.h:149]: Please add description for this field for consistency.

File: tests/data_files/Makefile
Comment (mprse) [inline tests/data_files/Makefile:302]: This test case covers failure of `mbedtls_x509_get_name`?

**Review by mprse — COMMENTED** [review]
LGTM!

Left only minor request and question.

File: include/mbedtls/x509_crt.h
Comment (AndrzejKurek) [reply include/mbedtls/x509_crt.h:149]: I'm not sure if this is beneficial. I would even remove the otherName description, as it doesn't introduce any new information :D

File: tests/data_files/Makefile
Comment (AndrzejKurek) [reply tests/data_files/Makefile:302]: Yes, this one is for `mbedtls_x509_get_name` failure, and the sequence tag malformation test is for  https://github.com/Mbed-TLS/mbedtls/pull/6900/files#diff-fe0507636920d9b43e948fb12d2850460e8cbe2b7d7c56278def07ea1ba41c6fR1853.

**Comment by AndrzejKurek** [comment]
I'll rebase this PR once https://github.com/Mbed-TLS/mbedtls/pull/6882 is merged, to not do the work twice.

**Comment by mprse** [comment]
This one can be rebase now and I think it should be merged next.

**Comment by AndrzejKurek** [comment]
Rebased. A new commit (last one) was added, since after the bugfix got merged, errors from parsing are no longer silently ignored.

**Comment by AndrzejKurek** [comment]
Rebased to fix conflicts with https://github.com/Mbed-TLS/mbedtls/pull/6997.

File: include/mbedtls/x509.h
Comment (mprse) [inline include/mbedtls/x509.h:298]: Why are we adding this here? This is part of https://github.com/Mbed-TLS/mbedtls/pull/7025.

Comment (AndrzejKurek) [reply include/mbedtls/x509.h:298]: What do you mean? Fixed whitespace?

Comment (mprse) [reply include/mbedtls/x509.h:298]: Oh sorry. It was too early. `unstructured_name` was already here and `directory_name` is inserted before.

**Review by mprse — DISMISSED** [review]

**Review by davidhorstmann-arm — CHANGES_REQUESTED** [review]
A couple of nits, looks good otherwise. I suppose you'll have to rebase on #7076 when that is merged.

**Comment by AndrzejKurek** [comment]
Internal CI failure: 

> java.io.IOException: Pipe is already closed
> Channel "unknown": Remote call on windows.eu-west-1a (i-072cbf6513a61d8c6) failed. The channel is closing down or has closed down

Seems like it's an infrastructure issue.

File: library/x509.c
Comment (AndrzejKurek) [reply library/x509.c]: Fixed, thanks!

File: include/mbedtls/x509.h
Comment (AndrzejKurek) [reply include/mbedtls/x509.h]: I'm not sure what's the problem with the current description, but I extended it a little bit - let me know what you think!

File: include/mbedtls/x509.h
Comment (davidhorstmann-arm) [inline include/mbedtls/x509.h:416]: Again, sorry to nitpick but this implies that we call `free()` directly on `san`, which is not true. We call free on any additional memory that `san` owns. I fear this may trip someone up in future if it's not clear.

This is the best wording I could come up with, but feel free to use a better one if you can find it:
```suggestion
 * \param san      SAN structure - extra memory owned by this structure will be freed
```

File: include/mbedtls/x509.h
Comment (davidhorstmann-arm) [inline include/mbedtls/x509.h:385]:
```suggestion
 *                 to dispose of this additional memory afterwards.
```
Just makes it clearer that we don't free the SAN struct itself.

**Review by davidhorstmann-arm — CHANGES_REQUESTED** [review]
Clarified nitpicks on the free-ing wording. Looks good otherwise.

**Review by mprse — APPROVED** [review]

File: tests/data_files/Makefile
Comment (d3zd3z) [inline tests/data_files/Makefile:350]: There is spurious whitespace on this line.

File: tests/data_files/Makefile
Comment (d3zd3z) [inline tests/data_files/Makefile:340]: I believe also whitespace here. Unfortunately github doesn't show it.

**Review by d3zd3z — CHANGES_REQUESTED** [review]
Just some really minor nits regarding whitespace.

File: include/mbedtls/x509.h
Comment (AndrzejKurek) [reply include/mbedtls/x509.h:416]: Ah, right, now I got it :)

**Review by davidhorstmann-arm — APPROVED** [review]
LGTM - thanks!

**Review by mprse — APPROVED** [review]
Changes looks good. Left only one suggestion from previous review.

File: include/mbedtls/x509.h
Comment (AndrzejKurek) [reply include/mbedtls/x509.h:297]: I'd like to address this in a follow-up PR that will cover some follow-up work for OPC-UA, if you're fine with that.

**Comment by AndrzejKurek** [comment]
API/ABI check complains about adding a union field, `mbedtls_x509_name directory_name;`.

**Comment by AndrzejKurek** [comment]
Adding a `DO-NOT-MERGE` label, as this is not crucial for this release, and might potentially result in some CI failures along with https://github.com/Mbed-TLS/mbedtls/pull/7009.

**Comment by mprse** [comment]
This one needs reabase, it was ready for merge before code freeze, so I think it can be merged after rebase.

**Comment by AndrzejKurek** [comment]
Rebased on the current development.
Conflicts in 3 commits:
- "Add a test for a malformed directoryname sequence"
- "Move an x509 malformation test"
- "Fix rfc822name test arguments"
All of the conflicts were in `test_suite_x509parse.data` due to switching from `MBEDTLS_HAS_ALG_SHA_256_VIA_MD_OR_PSA_BASED_ON_USE_PSA` to `MBEDTLS_MD_CAN_SHA256` on development - straightforward to fix.

**Review by mprse — APPROVED** [review]
Reviewed `test_suite_x509parse.data`.
LGTM

**Review by davidhorstmann-arm — APPROVED** [review]
LGTM - thanks!

(Reviewed the commits that had conflicts)

File: tests/data_files/test-ca.opensslconf
Comment (gilles-peskine-arm) [inline tests/data_files/test-ca.opensslconf:115]: Why `O=MALFORM_ME`? That doesn't seem to be part of the purpose of the `two-directorynames` test?

On the other hand, the content of `bad_second_directory_name_san` does seem related. And that section is unused. It looks like there's some kind of mistake here but I'm not sure what.

Comment (AndrzejKurek) [reply tests/data_files/test-ca.opensslconf:115]: Three files are using this extension:
- server5-two-directorynames.crt.der, which is a valid test;
- server5-directoryname-seq-malformed.crt.der, that's based on the valid file, but malforms the first directory name sequence tag;
- server5-second-directoryname-oid-malformed.crt.der, that's based on the valid file, but malforms the OID length of the second directoryName, containing "MALFORM_ME". It's just an indication that this directoryName will be malformed.

Comment (AndrzejKurek) [reply tests/data_files/test-ca.opensslconf:115]: In fact, `bad_second_directory_name_san` seems to be a leftover version of what I achieved with the malformation of the second directory name. `bad_second_directory_name_san` is unused and should be removed.

**Pull request merged**
